### PR TITLE
Errors & validations refactor

### PR DESCRIPTION
**Pull request by fede1608**
## Summary

- Errors refactor using next + error
- Validations removed from routes file

### CODE REVIEW

File: app/controllers/albums.js [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
+      .get('https://jsonplaceholder.typicode.com/albums')
+      .then(function(response) {
+        res.status(200).json({
+          albums: response.data.map(album => {
+            delete album.userId;
+            return album;
+          })
+        });
+      })
+      .catch(err => next(defaultError('External API error.')));
```
Comment (mpizzagalli) [inline app/controllers/albums.js]: maybe we can add the original error message from `err` also

File: app/controllers/albums.js [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
-      res.status(200).json({
-        albums: response.data.map(album => {
-          delete album.userId;
-          return album;
-        })
-      });
-    });
+    axios
+      .get('https://jsonplaceholder.typicode.com/albums')
+      .then(function(response) {
```
Comment (mpizzagalli) [inline app/controllers/albums.js]: use fat arrow functions

File: app/controllers/albums.js [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
-    axios.get('https://jsonplaceholder.typicode.com/albums').then(function(response) {
-      res.status(200).json({
-        albums: response.data.map(album => {
-          delete album.userId;
-          return album;
-        })
-      });
-    });
+    axios
+      .get('https://jsonplaceholder.typicode.com/albums')
```
Comment (mpizzagalli) [inline app/controllers/albums.js]: as I told you previously maybe we could move all the external api logic inside a service and abstract it there.
Also we should move the base url of the external api to a configurable value, so you can easily change between different environments like development, stage, qa or production without changing the code.

File: app/controllers/users.js [hunk 4]
```diff
@@ -1,56 +1,36 @@
...
-      return res.status(422).json({ errors: errors.array() });
-    }
     const { email, password } = req.body;
     User.scope('withPasswd')
       .findOne({ where: { email } })
       .then(user => {
         if (!user) {
           logger.error('Email is not registered.');
-          return res.status(401).json({ error: 'User auth failed. Check your email or password.' });
+          next(userUnauthorized);
```
Comment (mpizzagalli) [inline app/controllers/users.js]: this won't end the execution so you should add a return

File: app/routes.js [hunk 5]
```diff
@@ -1,60 +1,25 @@
...
+  app.post('/users/sessions', newSessionParamsValidation, userController.userNewSession);
   app.post('/users/sessions/invalidate_all', [tokenCheck], userController.invalidateAllSessions);
   app.get('/users', [tokenCheck], userController.usersList);
   app.get('/users/:id/albums', [tokenCheck], userController.albumList);
   app.get('/users/albums/:id/photos', [tokenCheck], userController.albumPhotosList);
-  app.post('/admin/users', [tokenCheck].concat(userValidations), userController.userAdminCreate);
-
+  app.post(
+    '/admin/users',
+    [tokenCheck, checkAdmin].concat(userParamsValidations),
```
Comment (mpizzagalli) [inline app/routes.js]: if you add it as a normal array element like `[tokenCheck, checkAdmin, userParamsValidations]` doesn't it work? As far as I remember express doesn't mind array of array of middlewares

**Review by mpizzagalli — CHANGES_REQUESTED** [review]
Nice work I left you some comments

File: app/controllers/albums.js (hunk 1 above)
Comment (fede1608) [reply app/controllers/albums.js]: Fixed

File: app/controllers/albums.js (hunk 3 above)
Comment (fede1608) [reply app/controllers/albums.js]: done in #14

File: app/controllers/albums.js (hunk 2 above)
Comment (fede1608) [reply app/controllers/albums.js]: Fixed in #14

File: app/routes.js (hunk 5 above)
Comment (fede1608) [reply app/routes.js]: Fixed

File: app/controllers/users.js (hunk 4 above)
Comment (fede1608) [reply app/controllers/users.js]: fixed

File: app/errors.js [hunk 6]
```diff
@@ -31,3 +31,7 @@ exports.invalidInput = internalError('Input invalid.', exports.INVALID_INPUT);
 
 exports.NEEDS_ADMIN = 'needs_admin';
 exports.needsAdmin = internalError('User is not admin.', exports.NEEDS_ADMIN);
+
+exports.EXTERNAL_API_ERROR = 'external_api_error';
+exports.externalApiError = error =>
+  internalError(error.message || 'External Api Error', exports.EXTERNAL_API_ERROR);
```
Comment (mpizzagalli) [inline app/errors.js:37]: As a comment you could use string interpolation and have
```
`External Api Error ${error.message}`
```

**Review by mpizzagalli — APPROVED** [review]

**Pull request merged**